domain/application/state/storage.go

```diff
@@ -110,12 +379,98 @@ WHERE  storage_id = $storageInstance.storage_id
+// - [github.com/juju/juju/domain/application/errors.StorageNameNotSupported]: when storage name is not defined in charm metadata.
+// - [github.com/juju/juju/domain/application/errors.InvalidStorageCount]: when the allowed attachment count would be violated.
+// - [github.com/juju/juju/domain/application/errors.InvalidStorageMountPoint]: when the filesystem being attached to the unit's machine has a mount point path conflict.
 func (st *State) AttachStorage(ctx context.Context, storageUUID corestorage.UUID, unitUUID coreunit.UUID) error {
-	//TODO implement me
-	return errors.New("not implemented")
+	db, err := st.DB()
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	countQuery, err := st.Prepare(`
+SELECT count(*) AS &storageCount.count
+FROM storage_instance si
+JOIN storage_unit_owner suo ON si.uuid = suo.storage_instance_uuid
+WHERE suo.unit_uuid = $storageCount.unit_uuid
+AND si.storage_name = $storageCount.storage_name
+AND si.uuid != $storageCount.uuid
+`, storageCount{})
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		// First to the basic life checks for the unit and storage.
+		unit, err := st.getMinimalUnit(ctx, tx, unitUUID)
+		if err != nil {
+			return err
+		}
+		if unit.LifeID != life.Alive {
+			return errors.Errorf("unit %q is not alive", unitUUID).Add(applicationerrors.UnitNotAlive)
+		}
+
+		stor, err := st.getStorageDetails(ctx, tx, storageUUID)
+		if err != nil {
+			return err
+		}
+		if stor.LifeID != life.Alive {
+			return errors.Errorf("storage %q is not alive", unitUUID).Add(applicationerrors.StorageNotAlive)
+		}
+
+		// See if the storage name is supported by the unit's current charm.
+		// We might be attaching a storage instance created for a previous charm version
+		// and no longer supported.
+		charmStorage, err := st.getUnitCharmStorageByName(ctx, tx, unitUUID, stor.StorageName)
+		if err != nil {
+			if errors.Is(err, sqlair.ErrNoRows) {
+				return errors.Errorf(
+					"charm for unit %q has no storage called %q",
+					unitUUID, stor.StorageName,
+				).Add(applicationerrors.StorageNameNotSupported)
+			}
+			return errors.Errorf("getting charm storage metadata for storage name %q unit %q: %w", stor.StorageName, unitUUID, err)
+		}
+
+		// Check allowed storage attachment counts - will this attachment exceed the max allowed.
+		// First get the number of storage instances (excluding the one we are attaching)
+		// of the same name already owned by this unit.
+		storageCount := storageCount{
+			StorageUUID: storageUUID,
+			StorageName: stor.StorageName,
+			UnitUUID:    unitUUID,
+		}
+		var count uint64
+		err = tx.Query(ctx, countQuery, storageCount).Get(&storageCount)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Errorf("querying storage count for storage %q on unit %q: %w", stor.StorageName, unitUUID, err)
+		}
+		if err == nil {
+			count = storageCount.Count
+		}
```

Comment:
Then this can go away.

Reply:
Actually, we don't need the count var anyway